packages/backend-common/src/database/types.ts

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import knex from 'knex';
+
+/**
+ * The PluginDatabaseClientFactory is used to provide a mechanism for backend plugins to obtain database connections
+ * for itself.
+ *
+ * The purpose of this factory is to allow plugins to get isolated data stores so that plugins are discouraged from
+ * database integration. Plugins can omit the `database` parameter to get the default plugin database, or
+ * provide an identifier that will be used to identify a separate database from the default.
+ */
+export type PluginDatabaseClientFactory = (database?: string) => Promise<knex>;
```

Comment:
I think it might be a bit cleaner to combine the manager and factory into a single type. So this could look like something like this:

```ts
export interface PluginDatabaseManager {
  getDatabaseClient(pluginId?: string) => Promise<knex>;
}
```

Reply:
Not sure what you mean by this comment. Didn't we move away from this design originally? 😅

I thought we decided on the functional type definition so that we can swizzle out any kind of implementation for plugins (for example, see the auth-backend standaloneServer.) Of course we can still swizzle things out by writing `{ getClient() { return yadda; } }`), but I think this is really preference and I'll change this to whatever the team thinks is conventional.

Reply:
Addressed in https://github.com/spotify/backstage/pull/2697/commits/35c0294018b8433787a795b7d9fcb6e9daef2b68.